airbyte-cdk/bulk/toolkits/load-db/src/main/kotlin/io/airbyte/cdk/load/orchestration/db/legacy_typing_deduping/TableNameMapFactory.kt

```diff
@@ -32,23 +38,185 @@ class TableCatalogFactory(
 ) {
     @Singleton
     fun get(): TableCatalog {
-        // TODO handle collisions in table names
-        return TableCatalog(
-            catalog.streams.associateWith { stream ->
+        if (catalog.streams.isEmpty()) {
+            throw ConfigErrorException(
+                "The catalog contained no streams. This likely indicates a platform/configuration error."
+            )
+        }
+
+        val processedRawTableNames = mutableSetOf<TableName>()
+        val processedFinalTableNames = mutableSetOf<TableName>()
+
+        val result = mutableMapOf<DestinationStream, TableNameInfo>()
+
+        catalog.streams.forEach { stream ->
+            val originalRawTableName = rawTableNameGenerator.getTableName(stream.descriptor)
+            val originalFinalTableName = finalTableNameGenerator.getTableName(stream.descriptor)
+            val currentRawProcessedName: TableName
+            val currentFinalProcessedName: TableName
+
+            // Create a hash-suffixed name to avoid collision
+            val hash =
+                DigestUtils.sha1Hex(
+                        "${originalFinalTableName.namespace}&airbyte&${stream.descriptor.name}"
+                    )
+                    .substring(0, 3)
+            val newName = "${stream.descriptor.name}_$hash"
+
+            if (originalRawTableName in processedRawTableNames) {
+                currentRawProcessedName = TableName(originalRawTableName.namespace, newName)
```

Comment:
we need to run newName through the table name generator again, otherwise we won't strip special characters

(... that probably deserves a test case)